src/main/native/cpp/analysis/FilteringUtils.cpp

```diff
@@ -91,3 +91,44 @@ units::second_t sysid::GetMeanTimeDelta(const Storage& data) {
 
   return std::accumulate(dts.begin(), dts.end(), 0_s) / dts.size();
 }
+
+void sysid::TrimQuasistaticData(std::vector<PreparedData>* data,
+                                double motionThreshold) {
+  data->erase(std::remove_if(data->begin(), data->end(),
+                             [motionThreshold](const auto& pt) {
+                               return std::abs(pt.voltage) <= 0 ||
+                                      std::abs(pt.velocity) < motionThreshold;
+                             }),
+              data->end());
+}
+
+void sysid::ApplyMedianFilter(std::vector<PreparedData>* data, int window) {
+  size_t step = window / 2;
+  frc::MedianFilter<double> medianFilter(window);
+  for (size_t i = 0; i < data->size(); i++) {
+    double median = medianFilter.Calculate(data->at(i).velocity);
+    if (i >= step) {
+      data->at(i - step).velocity = median;
+    }
+  }
+
+  // Starting points with the first value calculated by the median filter (since
+  // accel should be zero)
+  for (int i = 0; i < step; i++) {
+    data->at(i).velocity = data->at(step).velocity;
+  }
+
+  // Do the same procedure for the end datapoints
+  for (int i = data->size() - step; i < data->size(); i++) {
+    data->at(i).velocity = data->at(data->size() - step - 1).velocity;
+  }
```

Comment:
This would be more accurate if you include the padded values in the actual filter calculation. The boundary condition is that the acceleration is zero at the first point, not that the acceleration is zero for half the window period.

Reply:
So basically continue running the median filter by running Calculate with either the first or last value?

Reply:
Yea.

Reply:
Is the updated implementation what you were going for?

Reply:
Not quite. See my latest comment.